### ISSUE TITLE
[Feature Request] MPS provider

### ISSUE THREAD

**Issue by barakugav** [microsoft/onnxruntime#21271]
### Describe the feature request

Currently there is no MPS (apple Metal Performance Shaders GPU) provider. There is CoreML provider but from testing I did on my MacBook with arm64 M3 and MPS It does not use it.
Is there any plan to support MPS as a different provider or within CoreML?
I this there is a lot of value in such support, as a lot of the development is done MacBooks.

### Describe scenario use case

- Training small models on macos
- Testing new models on developer laptops before deploying

**Comment by skottmckay** [comment]
Did you check to see which nodes CoreML was actually assigned. And were you using NeuralNetwork or ML Program? 

If not, set log level to VERBOSE and look for 'Node placements' in the output.

Obviously resources aren't infinite and creating a new execution provider is a huge undertaking. CoreML is currently preferred as it offers access to GPU and NPU and we're increasing operator coverage for ML Program (as NeuralNetwork is deprecated).

A general-purpose GPU EP based on Vulkan (with Metal support via MoltenVK) may be feasible as it would run on multiple platforms. But the same condition will apply - the EP needs to support enough operators in the model to be effective.

**Comment by Rikyf3** [comment]
@skottmckay could you provide more details on the roadmap for CoreML EP?

Currently, the limited support for operators significantly slows down inference for most modern models. For example, operators like LayerNorm, GroupNorm, GeLU, and non-nearest neighbor resizing are not supported.

Additionally, many GitHub issues related to CoreML remain unanswered. While I understand the constraints on resources, it’s crucial for us developers to know whether ONNX Runtime with CoreML EP is a viable option or if we should consider alternatives. In my specific case, a [model](https://github.com/megvii-research/NAFNet) that takes one minute to run on DirectML EP takes 50 minutes on CoreML.

**Comment by skottmckay** [comment]
We typically add operator support based on production use cases. If you have such a use case please provide details and the operators required. 

e.g. we recently added ML Program support for quite a few operators, including [Resize](https://github.com/microsoft/onnxruntime/pull/21370/) (although there are caveats with Resize as we're not able to map all the ONNX options to equivalent CoreML parameters, and CoreML seems to have inconsistent implementations of some operators): https://github.com/microsoft/onnxruntime/pulls?q=is%3Apr+coreml+is%3Aclosed

We would most likely limit new operator support to ML Program though as NeuralNetwork is deprecated. ML Program requires iOS 15+ or macOS 12+.

There may also be differences in what CoreML operators are available and what options they support vs. ONNX.

[GeLU](https://apple.github.io/coremltools/source/coremltools.converters.mil.mil.ops.defs.html#coremltools.converters.mil.mil.ops.defs.iOS15.activation.gelu) is supported, as is [LayerNorm](https://apple.github.io/coremltools/source/coremltools.converters.mil.mil.ops.defs.html#coremltools.converters.mil.mil.ops.defs.iOS15.normalization.layer_norm). GroupNorm may map to CoreML [batch_norm](https://apple.github.io/coremltools/source/coremltools.converters.mil.mil.ops.defs.html#coremltools.converters.mil.mil.ops.defs.iOS15.normalization.batch_norm).

**Comment by Rikyf3** [comment]
@skottmckay, thank you for your prompt response.

I’m currently using onnxruntime==1.18.1 on macOS, so it seems that the improvements you mentioned might be available in the upcoming version. I have implemented a U-Net model in PyTorch, and based on the verbosity log, I noticed that GeLU is decomposed into an Erf along with multiplications and sums, but Erf is not supported. Similarly, GroupNorm decomposes into Reshape + InstanceNorm + Reshape, and none of these operators are supported (Reshape for dynamic shape issue).

I will attach my unet implementation if you need it.
[unet.py.zip](https://github.com/user-attachments/files/16467876/unet.py.zip)

**Comment by divideconcept** [comment]
It'd be great is there was another mac alternative to CoreML, which unfortunately cannot run a lot of models... And is still limited to have tensor size dimension <=16384.
MPS would be my preferred choice as it does not have this 16384 limitation, and PyTorch seems to run all models I need with their MPS EP.
CoreML seems to be very low priority at Apple, they don't seem to care at all about improving this tech for years...

**Comment by interactda** [comment]
Please add MPS support thank you, onnx runtime runs 5 times slower on MPS than the original model run on MPS

**Comment by iskng** [comment]
microsoft/onnxruntime

**Comment by mxkrn** [comment]
I'm interested in hearing more about whether this is being considered for development.

We're building a `cpp` application that use `onnxruntime` for running model inference on device. Having MPS support would provide a massive performance boost which will lead massive UX improvements.

Ideally we'd like direct access to MPS as a hardware provider rather than having to take a detour via CoreML.

**Comment by barakugav** [comment]
[executroch](https://github.com/pytorch/executorch) has a support for MPS if someone is looking for alternatives.
The library is written in Cpp, but I wrote Rust bindings to it, check it out: https://github.com/barakugav/executorch-rs/

**Comment by B-dev-cyber** [comment]
Please implement this feature, mac using onnx models are IMPOSSIBLE to have high efficiency.

**Comment by hg0428** [comment]
CoreML does not support input dim > 16384. Input:model.embed_tokens.weight, shape: {151936,2048}

number of nodes supported by CoreML: 84 (out of 602)

Yeah, I need MPS.

**Comment by YM2132** [comment]
Does CoreML not use the metal processor?

**Comment by hg0428** [comment]
> Does CoreML not use the metal processor?

The difference is that CoreML also runs on the NPU, is less mature, and far more restrictive than an MPS backend.

**Comment by YM2132** [comment]
> > Does CoreML not use the metal processor?
> 
> The difference is that CoreML also runs on the NPU, is less mature, and far more restrictive than an MPS backend.

Do you know why was CoreML chosen as the execution provider for onnxruntime? It seems that a straight up MPS one would be much better. 

Do you know how big of a task implementing an MVP mps execution provider would be? Is anyone working on this at the moment?

**Comment by robertknight** [comment]
> The difference is that CoreML also runs on the NPU, is less mature, and far more restrictive than an MPS backend.

To add to the confusion, MPS doesn't use the NPU, but the similarly named [MPS Graph](https://developer.apple.com/documentation/metalperformanceshadersgraph) framework does. According to the transcript of [this WWDC session](https://developer.apple.com/videos/play/wwdc2024/10218/), it is what CoreML uses under the hood.

**Comment by YM2132** [comment]
@robertknight, wait what? MPS is the Apple equivalent to CUDA for the Metal GPU right? 

MPS does not use NPU, but the MPS Graph framework you mention can run on NPU (and thats why CoreML runs across CPU, MPS and ANE?)

**Comment by ChinChangYang** [comment]
An [MLX](https://github.com/ml-explore/mlx) executor is the future. MLX has been tested in many project, such as [mlx-lm](https://github.com/ml-explore/mlx-lm) for LLM. It exhibits performance in an equal level of MPS Graph, and provides pure C++ [APIs](https://ml-explore.github.io/mlx/build/html/cpp/ops.html). MPS or MPS Graph needs C++/Swift interoperability or Objective-C/C++ to work with a C++ project, but MLX only needs an MLX-installed environment. Moreover, MLX provides more [quantization](https://ml-explore.github.io/mlx/build/html/python/_autosummary/mlx.core.quantize.html#mlx.core.quantize) APIs that are much easier to be done than MPS and MPS Graph.

**Comment by robertknight** [comment]
> MPS does not use NPU, but the MPS Graph framework you mention can run on NPU (and thats why CoreML runs across CPU, MPS and ANE?)

That is my understanding. Unfortunately the NPU doesn't have a public low-level API like eg. Qualcomm's Hexagon does for example.

**Comment by justinchuby** [comment]
Work in progress: https://github.com/[justinchuby/onnxruntime-mps. The mlx suggestion is also a good one. Will try that next.

**Comment by divideconcept** [comment]
@justinchuby thanks for taking this task, we're very much looking forward to this !

**Comment by justinchuby** [comment]
Note that this is not committed work but rather a personal side project.

**Comment by justinchuby** [comment]
Try it here: https://pypi.org/project/onnxruntime-mlx/
